File: .thread_autoload_injector.php
<?php

function getPhpFiles(string $directory): array
{
	$rii = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($directory));
	$files = [];

	foreach ($rii as $file) {
		if ($file->isDir() || $file->getExtension() !== "php") {
			continue;
		}

		$files[] = $file->getPathname();
	}

	return $files;
}

function getClassFromFile(string $file): ?array
{
	$content = file_get_contents($file);
	if ($content === false) {
		echo "❌ Ошибка чтения файла $file\n";
		return null;
	}

	preg_match('/namespace\s+([^;]+);/', $content, $namespaceMatch);
	$namespace = $namespaceMatch[1] ?? '';

	preg_match('/class\s+(\w+)\s+extends\s+(\w+)/', $content, $classMatch);
	if ($classMatch) {
		[$fullMatch, $className, $parentClass] = $classMatch;
		$fullClass = $namespace ? "$namespace\\$className" : $className;
		return ['full_class' => $fullClass, 'parent_class' => $parentClass, 'file' => $file];
	}

	return null;
}

function modifyOnRunMethod(string $file, string $autoloadPath): void
{
	$content = file_get_contents($file);
	if ($content === false) {
		echo "❌ Ошибка чтения файла $file\n";
		return;
	}

	if (strpos($content, 'require_once') !== false) {
		echo "❌ Пропущено (require_once уже присутствует): $file\n";
		return;
	}

	if (preg_match('/(?:public|protected)?\s*function\s+onRun\s*\(\s*\)\s*(?::\s*\w+)?\s*{?/', $content, $matches, PREG_OFFSET_CAPTURE)) {
		$insertPos = $matches[0][1] + strlen($matches[0][0]);
		$newContent = substr_replace($content, "\n\t\trequire_once __DIR__ . '/$autoloadPath';\n", $insertPos, 0);

		if (file_put_contents($file, $newContent) !== false) {
			echo "✅ Изменен файл: $file\n";
		} else {
			echo "❌ Ошибка записи в файл $file\n";
		}
	} else {
		echo "⚠️ Не найден метод onRun в файле: $file\n";
	}
}

$pluginPath = __DIR__;
$autoloadPath = "autoload.php";
$threadClasses = ["Thread", "AsyncTask"];
$ignorePaths = ["vendor"];
$extraUp = ["virions" => 1];

$phpFiles = getPhpFiles($pluginPath);
echo "🔍 Найдено " . count($phpFiles) . " PHP файлов.\n";

foreach ($phpFiles as $file) {
	$classData = getClassFromFile($file);
	if ($classData && in_array($classData['parent_class'], $threadClasses, true)) {
		if (array_intersect(explode(DIRECTORY_SEPARATOR, $file), $ignorePaths)) {
			echo "🚫 Пропущен (игнорируемая папка): $file\n";
			continue;
		}

		$namespaceParts = explode("\\", $classData["full_class"]);
		$extraUpLevels = array_sum(array_intersect_key($extraUp, array_flip(explode(DIRECTORY_SEPARATOR, $file))));
		$levelsUp = count($namespaceParts) + $extraUpLevels;
		$relativeAutoloadPath = str_repeat("../", (int) $levelsUp) . $autoloadPath;

		modifyOnRunMethod($file, $relativeAutoloadPath);
	}
}

echo "🎉 Готово!\n";

File: .thread_autoload_injector.py
"""
Мне легче работать в Python, так что не вам меня судить
"""

import re
from pathlib import Path

# Пути
PLUGIN_PATH = Path(__file__).parent
AUTOLOAD_PATH = "autoload.php"
THREAD_CLASSES = [
	"Thread", "AsyncTask"
]
IGNORE_PATHES = [
	"vendor"
]
EXTRA_UP = {
	"virions": 1
}

def get_php_files(directory: Path):
	"""Получает список всех PHP файлов в указанной директории."""
	return list(directory.rglob("*.php"))

def get_class_from_file(file: Path):
	"""Извлекает информацию о классе из PHP файла."""
	try:
		content = file.read_text(encoding="utf-8")
	except Exception as e:
		print(f"❌ Ошибка чтения файла {file}: {e}")
		return None

	namespace_match = re.search(r'namespace\s+([^;]+);', content)
	namespace = namespace_match.group(1).strip() if namespace_match else ''

	class_match = re.search(r'class\s+(\w+)\s+extends\s+(\w+)', content)

	if class_match:
		class_name, parent_class = class_match.groups()
		full_class = f"{namespace}.{class_name}" if namespace else class_name

		return {
			"full_class": full_class,
			"parent_class": parent_class,
			"file": file
		}

	return None

def modify_on_run_method(file: Path, autoload_path: str):
	"""Добавляет require_once в метод onRun, если его еще нет."""
	try:
		content = file.read_text(encoding="utf-8")
	except Exception as e:
		print(f"❌ Ошибка чтения файла {file}: {e}")
		return

	if "require_once" in content:
		print(f"❌ Пропущено (require_once уже присутствует): {file}")
		return

	match = re.search(r'(?:public|protected)?\s*function\s+onRun\s*\(\s*\)\s*(?::\s*\w+)?\s*{?', content, re.MULTILINE)

	if match:
		insert_pos = match.end()
		new_content = content[:insert_pos] + f'\n\t\trequire_once __DIR__ . "/{autoload_path}";\n' + content[insert_pos:]

		try:
			file.write_text(new_content, encoding="utf-8")
			print(f"✅ Изменен файл: {file}")
		except Exception as e:
			print(f"❌ Ошибка записи в файл {file}: {e}")
	else:
		print(f"⚠️ Не найден метод onRun в файле: {file}")

# Обработка файлов
php_files = get_php_files(PLUGIN_PATH)
print(f"🔍 Найдено {len(php_files)} PHP файлов.")

for file in php_files:
	class_data = get_class_from_file(file)

	if class_data and class_data["parent_class"] in THREAD_CLASSES:
		# Пропустить файлы, если путь содержит одну из игнорируемых папок
		if any(ignored in file.parts for ignored in IGNORE_PATHES):
			print(f"🚫 Пропущен (игнорируемая папка): {file}")
			continue

		namespace_parts = class_data["full_class"].split("\\")

		# Вычисляем дополнительные уровни up на основе всех папок из EXTRA_UP
		extra_up = sum(EXTRA_UP.get(part, 0) for part in file.parts)

		levels_up = len(namespace_parts) + extra_up
		autoload_path = ("../" * levels_up) + AUTOLOAD_PATH

		modify_on_run_method(file, autoload_path)

print("🎉 Готово!")
